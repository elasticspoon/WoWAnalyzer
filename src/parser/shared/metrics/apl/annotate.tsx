import { SpellLink } from 'interface';
import React from 'react';

import type { CheckResult, Rule } from './index';

export function RuleDescription({ rule }: { rule: Rule }) {
  if (!('condition' in rule)) {
    return null;
  }

  return (
    <>
      The <SpellLink id={rule.spell.id} /> rule was active because {rule.condition.describe()}.
    </>
  );
}

/**
 * Annotate a list of APL violations on the timeline.
 *
 * This mutates the input and will overwrite other annotations, but is
 * idempotent. There is no return value.
 */
export default function annotateTimeline(violations: CheckResult['violations']) {
  for (const violation of violations) {
    violation.actualCast.meta = {
      isInefficientCast: true,
      inefficientCastReason: (
        <>
          <SpellLink id={violation.expectedCast.id} /> was available and higher priority.
          <RuleDescription rule={violation.rule} />
        </>
      ),
    };
  }
}
