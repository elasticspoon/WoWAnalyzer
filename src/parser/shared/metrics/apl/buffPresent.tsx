import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import { EventType } from 'parser/core/Events';
import React from 'react';

import { Condition, tenseAlt } from './index';

export interface PandemicData {
  timeRemaining: number;
  duration: number;
  pandemicCap?: number;
}

export function buffPresent(spell: Spell): Condition<boolean> {
  return {
    key: `buffPresent-${spell.id}`,
    init: () => false,
    update: (state, event) => {
      switch (event.type) {
        case EventType.ApplyBuff:
          if (event.ability.guid === spell.id) {
            return true;
          }
          break;
        case EventType.RemoveBuff:
          if (event.ability.guid === spell.id) {
            return false;
          }
          break;
      }

      return state;
    },
    validate: (state, _event) => state,
    describe: (tense) => (
      <>
        <SpellLink id={spell.id} /> {tenseAlt(tense, 'is', 'was')} present
      </>
    ),
  };
}

interface DurationData {
  timeRemaining: number;
  referenceTime: number;
}

const buffDuration = (timeRemaining: number | undefined, pandemic: PandemicData): number =>
  Math.min(
    (timeRemaining || 0) + pandemic.duration,
    pandemic.duration * (pandemic.pandemicCap || 3),
  );

export function buffMissing(
  spell: Spell,
  optPandemic?: PandemicData,
): Condition<DurationData | null> {
  const pandemic: PandemicData = {
    timeRemaining: 0,
    duration: 0,
    ...optPandemic,
  };

  return {
    key: `buffMissing-${spell.id}`,
    init: () => null,
    update: (state, event) => {
      switch (event.type) {
        case EventType.ApplyBuff:
          if (event.ability.guid === spell.id) {
            return {
              referenceTime: event.timestamp,
              timeRemaining: buffDuration(state?.timeRemaining, pandemic),
            };
          }
          break;
        case EventType.RemoveBuff:
          if (event.ability.guid === spell.id) {
            return null;
          }
          break;
      }

      return state;
    },
    validate: (state, event) =>
      state === null ||
      state.referenceTime + state.timeRemaining < event.timestamp + pandemic.timeRemaining,
    describe: (tense) => (
      <>
        <SpellLink id={spell.id} /> {tenseAlt(tense, 'is', 'was')} missing{' '}
        {optPandemic && <>or about to expire</>}
      </>
    ),
  };
}
