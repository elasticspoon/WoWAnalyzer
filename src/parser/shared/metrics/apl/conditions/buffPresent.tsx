import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import { EventType } from 'parser/core/Events';
import React from 'react';

import { Condition, tenseAlt } from '../index';

export interface PandemicData {
  timeRemaining: number;
  duration: number;
  pandemicCap?: number;
}

export function buffPresent(spell: Spell): Condition<boolean> {
  return {
    key: `buffPresent-${spell.id}`,
    init: () => false,
    update: (state, event) => {
      switch (event.type) {
        case EventType.ApplyBuff:
          if (event.ability.guid === spell.id) {
            return true;
          }
          break;
        case EventType.RemoveBuff:
          if (event.ability.guid === spell.id) {
            return false;
          }
          break;
      }

      return state;
    },
    validate: (state, _event) => state,
    describe: (tense) => (
      <>
        <SpellLink id={spell.id} /> {tenseAlt(tense, 'is', 'was')} present
      </>
    ),
  };
}

interface DurationData {
  timeRemaining: number;
  referenceTime: number;
}

const buffDuration = (timeRemaining: number | undefined, pandemic: PandemicData): number =>
  Math.min(
    (timeRemaining || 0) + pandemic.duration,
    pandemic.duration * (pandemic.pandemicCap || 3),
  );

/**
   The rule applies when the buff `spell` is missing. The `optPandemic`
   parameter gives the ability to allow early refreshes to prevent a buff
   dropping, but this will not *require* early refreshes.
 **/
export function buffMissing(
  spell: Spell,
  optPandemic?: PandemicData,
): Condition<DurationData | null> {
  const pandemic: PandemicData = {
    timeRemaining: 0,
    duration: 0,
    ...optPandemic,
  };

  return {
    key: `buffMissing-${spell.id}`,
    init: () => null,
    update: (state, event) => {
      switch (event.type) {
        case EventType.ApplyBuff:
          if (event.ability.guid === spell.id) {
            return {
              referenceTime: event.timestamp,
              timeRemaining: buffDuration(state?.timeRemaining, pandemic),
            };
          }
          break;
        case EventType.RemoveBuff:
          if (event.ability.guid === spell.id) {
            return null;
          }
          break;
      }

      return state;
    },
    validate: (state, event, ruleSpell) => {
      if (state === null) {
        // buff is missing
        return true;
      } else if (state.referenceTime + 200 > event.timestamp) {
        // buff was *just* applied, possibly by this very spell. treat it as optional
        return event.ability.guid === ruleSpell.id;
      } else {
        // otherwise, return true if we can pandemic this buff
        return (
          ruleSpell.id === event.ability.guid &&
          state.referenceTime + state.timeRemaining < event.timestamp + pandemic.timeRemaining
        );
      }
    },
    describe: (tense) => (
      <>
        <SpellLink id={spell.id} /> {tenseAlt(tense, 'is', 'was')} missing{' '}
        {optPandemic && <>or about to expire</>}
      </>
    ),
  };
}
