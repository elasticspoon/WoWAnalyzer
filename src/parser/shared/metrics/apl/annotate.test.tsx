import React from 'react';
import renderer from 'react-test-renderer';

import { ConditionDescription } from './annotate';
import { buffPresent } from './buffPresent';

describe('ConditionDescription', () => {
  it('should return no description for unconditional rules', () => {
    const content = renderer
      .create(<ConditionDescription rule={{ id: 1, name: 'Test', icon: '' }} />)
      .toJSON();
    expect(content).toBeNull();
  });

  it('should return a description for a rule with a condition', () => {
    const rule = {
      spell: { id: 1, name: 'Test', icon: '' },
      condition: buffPresent({ id: 2, name: 'Buff', icon: '' }),
    };

    const content = renderer.create(<ConditionDescription rule={rule} />).toJSON();

    expect(content).toMatchSnapshot();
  });
});
