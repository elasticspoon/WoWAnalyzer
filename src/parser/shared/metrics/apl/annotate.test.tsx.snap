// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionDescription should return a description for a rule with a condition 1`] = `
Array [
  " ",
  "because",
  " ",
  <a
    className="spell-link-text"
    href="http://wowhead.com/spell=2"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt=""
      className="icon game "
      src="//render-us.worldofwarcraft.com/icons/56/inv_misc_questionmark.jpg"
    />
     
    Unknown spell: 2
  </a>,
  " ",
  "was",
  " present",
]
`;
